accepted/00000-upgrading-bootstrap.md

```diff
@@ -0,0 +1,70 @@
+
+Coming out successfully from the other side, we would hope to have simpler stylesheets with less to maintain, and fewer security issues.  
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Theme & styles
+
+Bootstrap 3 uses Less while Bootstrap 5 only offers Sass styles. This switch by itself is not very problematic, however there are additional issues that make the upgrade more difficult. Our current styles make heave use of numerous mixins from Bootstrap 3 which no longer exist in Bootstrap 5. Additionally, there are numerous breaking styling changes going from 3 to 4 and even more so moving from 4 to 5.  
+
+Instead of trying to manually migrate our existing very large theme source, I propose we instead try to first reduce our existing theme footprint as much as possible. We already have component scoped styles working and in use, so we can try move component or view specific styles (e.g. table styles, setup wizard etc) into component scoped styles and remove them from the global stylesheet.  
+
+This allows us to cut up the effort required and do at least some of the work gradually. I suspect the last core theme change will still be painful, though. It might be easier to rebuild the existing theme from the base Bootstrap style rather than migrating the existing one, but this is something we can consider after we've simplified the theme.  
+
+## Component and class name changes
+
+This is tied to the previous section, but worth discussing on its own. Numerous component, layout and utility classes have been dropped or changed in both upgrades, e.g. `input-group-addon` was changed in Bootstrap 4 and dropped in 5, so we need to use `input-group-text` and check if it renders correctly etc. The list is pretty long. As discussed in [this ticket](https://github.com/SUSE/spacewalk/issues/18346), it's unlikely that we'd find a tool that can handle all different ways we use Bootstrap (in JSP, JPSF, React, etc), however writing a migration script similar to the Typescript migration might be feasible. For this, we can assemble a list that covers all the breaking changes that affect us by going through the list of changes for 4 and 5, checking whether we use the affected classes, and if so, write down what the upgrade path is.  
+
+## Scripts
+
+There are some scripting changes, e.g. static method names have changed `_foo()` → `foo()`, data properties have changed `data-foo` → `data-bs-foo` etc. This mostly concerns old pages and I would expect most issues here to be straightforward to fix once we get to a state where we can run the test suite on it.
+
+## Summary
+
+In short, the suggested upgrade path can be summarized as follows:  
+
+ - Simplify the existing styles, trying to use standard Bootstrap tooling wherever possible
+ - Gather the list of breaking changes
+ - Write a migration script and/or manually migrate the above list
+ - Migrate existing styles from Less to Sass
+ - Verify the results with Cucumber
```

Comment:
At that point, it will not be only verifying but also adapting the test framework code to use the new WebElement locators.

Reply:
I would definitely encourage to follow best practices when writing new code or changing existing code in a significant way, but I do not think adapting the whole testsuite to a new style of locators really belongs to this RFC (or effort).

In principle it could be a separate one, correct?

Reply:
Yes it could be in theory, however there is absolutely no benefit to it except tech-debt. Since the locators then need to be rewritten again after the PR with the work this RFC outlines (because of the breaking bootstrap changes), this would be a duplicated effort imho.